tools/validators/instance_validator/instance_parser.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the License);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Parses and validates YAML instance files for syntax"""
+
+import strictyaml as syaml
+
+# TODO check valid ontological content in next validation steps
+"""strictyaml schema parses a YAML instance from its first level of keys
+github.com/google/digitalbuildings/blob/master/ontology/docs/building_config.md#config-format"""
+_SCHEMA = syaml.MapPattern(syaml.Str(),
+                           syaml.Map({
+                               'type': syaml.Str(),
+                               'id': syaml.Str(),
+                               syaml.Optional('connections'):
+                               syaml.MapPattern(syaml.Str(),
+                                                syaml.Str()) |
+                               syaml.Seq(
+                                   syaml.MapPattern(syaml.Str(),
+                                                    syaml.Str())),
+                               syaml.Optional('links'): syaml.MapPattern(
+                                   syaml.Str(),
+                                   syaml.MapPattern(syaml.Str(), syaml.Str())),
+                               syaml.Optional('translation'): syaml.Any(),
+                               syaml.Optional('metadata'): syaml.Any()
+                               }))
+
+"""schema separately parses translation to account for multiple valid formats
+github.com/google/digitalbuildings/blob/master/ontology/docs/building_config.md#defining-translations"""
+_TRANSLATION_SCHEMA = syaml.Str() | syaml.Any()
+
+# TODO add manual check to de-duplicate units/unit_values/states
+# TODO add all units/unit_values/states to translation_data_schema
+"""further account for multiple valid translation formats
+github.com/google/digitalbuildings/blob/master/ontology/docs/building_config.md#defining-translations"""
+_TRANSLATION_DATA_SCHEMA = syaml.Str() | syaml.Map({
+    'present_value': syaml.Str(),
+    syaml.Optional('states'): syaml.MapPattern(syaml.Str(), syaml.Str()),
+    syaml.Optional('units'): syaml.Map({
+        'key': syaml.Str(),
+        'values': syaml.MapPattern(syaml.Str(), syaml.Str())
+    }),
+    syaml.Optional('unit_values'): syaml.MapPattern(syaml.Str(), syaml.Str())
+    })
+
+_COMPLIANT = 'COMPLIANT'
```

Comment:
this is confusing here, move to the top of the file?

Reply:
Done